app/code/Magento/AmqpStore/Test/Unit/Plugin/AsynchronousOperations/MassConsumerEnvelopeCallbackTest.php

```diff
@@ -0,0 +1,200 @@
+
+        $message = 'no_such_entity_exception';
+        $this->storeManagerMock
+            ->expects($this->once())
+            ->method('getStore')
+            ->willThrowException(new NoSuchEntityException(__($message)));
+
+        $this->loggerMock
+            ->expects($this->once())
+            ->method('error')
+            ->with("Can't set currentStoreId during processing queue. Message rejected. Error $message.");
+
+        $queue = $this->getMockBuilder(QueueInterface::class)
+            ->disableOriginalConstructor()
+            ->getMockForAbstractClass();
+        $this->subjectMassConsumerEnvelopeCallbackMock
+            ->expects($this->once())
+            ->method('getQueue')
+            ->willReturn($queue);
+        $queue
+            ->expects($this->once())
+            ->method('reject')
+            ->with($this->messageMock, false, $message);
+
+        $isProceedCalled = false;
+        $proceed = function ($message) use (&$isProceedCalled) {
+            $isProceedCalled = !!$message;
+        };
+
+        $this->massConsumerEnvelopeCallbackPlugin->aroundExecute(
+            $this->subjectMassConsumerEnvelopeCallbackMock,
+            $proceed,
+            $this->messageMock
+        );
+
+        $this->assertFalse($isProceedCalled);
+    }
+
+    /**
+     * @dataProvider provideApplicationHeadersForAroundExecuteSuccess
+     *
+     * @param mixed $headers
+     * @param int $storeId
+     * @param int $currentStoreId
+     */
+    public function testAroundExecuteWhenSuccess($headers, int $storeId, int $currentStoreId)
+    {
+        $amqpProperties = ['application_headers' => $headers];
+        $this->messageMock->expects($this->once())
+            ->method('getProperties')
+            ->willReturn($amqpProperties);
+
+        $this->storeManagerMock->expects($this->once())
+            ->method('getStore')
+            ->willReturn($this->storeMock);
+        $this->storeMock->expects($this->once())
+            ->method('getId')
+            ->willReturn($currentStoreId);
+
+        $this->storeManagerMock->expects($this->exactly(2))
+            ->method('setCurrentStore')
+            ->withConsecutive(
+                [$storeId],
+                [$currentStoreId]
+            );
+
+        $isProceedCalled = false;
+        $proceed = function ($message) use (&$isProceedCalled) {
+            $isProceedCalled = !!$message;
+        };
+
+        $this->massConsumerEnvelopeCallbackPlugin->aroundExecute(
+            $this->subjectMassConsumerEnvelopeCallbackMock,
+            $proceed,
+            $this->messageMock
+        );
+
+        $this->assertTrue($isProceedCalled);
+    }
+
+    public function provideApplicationHeadersForAroundExecuteSuccess()
+    {
+        $storeId = 123;
+        $currentStoreId = 99;
+
+        return [
+            [['store_id' => 123], $storeId, $currentStoreId],
+            [new AMQPTable(['store_id' => 123]), $storeId, $currentStoreId],
```

Comment:
I created a new object AMQPTable because I could not create a mock and mock for the final method "getNativeData"

Do you have any suggestions for this?

https://github.com/mbvb1223/magento2/blob/1e675494b29424bed5d2183e85b4f619cf6e4c47/app/code/Magento/AmqpStore/Plugin/AsynchronousOperations/MassConsumerEnvelopeCallback.php#L75